framework/src/st_things/things_stack/src/stack/things_stack.c

```diff
@@ -0,0 +1,1149 @@
+		things_stop();
+		return 0;
+	}
+	// [JAY] Will be refactored to received callbacks for GET/PUT/POST/DELETE/OBSERVE ..
+	g_req_handler->init_module();
+
+	// 4. Initiate Server Builder
+	g_server_builder = get_builder_instance();
+	if (NULL == g_server_builder) {
+		THINGS_LOG_ERROR(THINGS_ERROR, TAG, "Failed to initialize Resource Server");
+		things_stop();
+		return 0;
+	}
+
+	g_server_builder->init_module(g_server_builder, g_req_handler->entity_handler);
+
+#ifdef __SECURED__
+	//5. Generate MAC based device UUID
+	if (0 != sm_generate_mac_based_device_id(false)) {
+		THINGS_LOG(THINGS_WARNING, TAG, "Failed to generate MAC based device_id");
+		things_stop();
+		return 0;
+	}
+#endif
+
+	// 6. Register Device-ID & Resources
+	if (dm_register_device_id() == false) {
+		THINGS_LOG_ERROR(THINGS_ERROR, TAG, "Failed to register Device ID");
+		things_stop();
+		return 0;
+	}
+
+	if (!dm_register_resource(g_server_builder)) {
+		THINGS_LOG_ERROR(THINGS_ERROR, TAG, "Failed to register Resource");
+		things_stop();
+		return 0;
+	}
+	// 7. Initiate Easy-Setup & Login to Cloud
+	if (ESM_OK != esm_init_easysetup(0, g_server_builder)) {
+		THINGS_LOG_ERROR(THINGS_ERROR, TAG, "Failed to initialize Easy-Setup Module");
+		things_stop();
+		return 0;
+	}
+	// Register Callback
+	esm_register_cloud_cb(cloud_data_cb_esm);
+
+	if (es_cloud_init(g_server_builder) == NULL) {
+		THINGS_LOG_ERROR(THINGS_ERROR, TAG, "Failed to initialize Cloud");
+		things_stop();
+		return 0;
+	}
+
+	is_things_module_inited = 1;
+
+	g_quit_flag = 0;
+
+	THINGS_LOG_D(THINGS_INFO, TAG, "Stack Initialization Success");
+	return 1;
+}
+
+int things_reset(void *remote_owner, things_es_enrollee_reset_e resetType)
+{
+	THINGS_LOG_D(THINGS_DEBUG, TAG, THINGS_FUNC_ENTRY);
+
+	if (resetType == RST_AUTO_RESET) {
+		try_turn_on_soft_ap();
+	}
+
+	int res = -1;
+
+	if (b_thread_things_reset) {
+		return 0;
+	}
+
+	if (resetType >= RST_ENUM_EOF || resetType < RST_NEED_CONFIRM) {
+		THINGS_LOG_D(THINGS_INFO, TAG, "Not support things_es_enrollee_reset_e value(%d). So, Set value with RST_NEED_CONFIRM", resetType);
+		resetType = RST_NEED_CONFIRM;
+	}
+
+	if (g_confirm_reset_start == NULL) {
+		THINGS_LOG_ERROR(THINGS_ERROR, TAG, "Reset-Start confirmation Function is not registered. It's mandatory");
+		return res;
+	}
+
+	pthread_mutex_lock(&m_thread_oic_reset);
+	if (b_thread_things_reset == false) {
+		b_thread_things_reset = true;
+		reset_args_s *args = (reset_args_s *) things_malloc(sizeof(reset_args_s));
+		if (args == NULL) {
+			THINGS_LOG_ERROR(THINGS_ERROR, TAG, "Failed to allocate reset_args_s memory");
+			b_thread_things_reset = false;
+			res = -1;
+			goto GOTO_OUT;
+		}
+		args->remote_owner = (things_resource_s *) remote_owner;
+		args->resetType = resetType;
+
+		b_reset_continue_flag = true;
+#ifdef __ST_THINGS_RTOS__
+		if (pthread_create_rtos(&h_thread_things_reset, NULL, (pthread_func_type) t_things_reset_loop, args, THINGS_STACK_RESETLOOP_THREAD) != 0)
```

Comment:
all of pthread don't have pthread_join or detach. That is a memory leak. You must fix it.

Reply:
modified - h_thread_things_wifi_join / h_thread_things_reset
In case of 'h_thread_things_reset' thread, 'pthread_join' is called when 'things_stop' is called. Eventhough that, is it necessaty to call 'pthread_detach' after pthread create?

Reply:
modified

Reply:
In things_stop(), we don't need to call join() because the h_thread_things_reset is already detached.

Reply:
modified